### PR TITLE
[WIP] Typing and parameterizing entity methods

### PR DESCRIPTION
**Pull request by daemitus**
### Proposed changes

* kwargs is evil, add parameters. Keep the keyword requirement by using a `*` parameter.
* Cleanup the method contents, less `if key in thing`, more `thing.get(key, default)`

### Related issues

* N/A

### Checklist

- [ ] I consider the submitted work as finished
- [ ] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments

One of the biggest problems I have is trying to figure out what parameters are allowed where without looking at the method bodies. So may as well try extracting parameters from where they are used, add typing based on the gql schema, and see how it looks after. 

One design consideration is utilizing the walrus operator, which would mean dropping py3.7 support. It does still have 10 months of life left in it before EOL. 

- Should import_from_stix2 be typed as `Union[Dict[str, any], stix2.SomeStixType]`?
- Are there other code patterns that are preferred? 
- Need to go back and turn on compatibility checks, just saw a typing `|` operator that's 3.10+. 
- Deprecating non-snake-case arguments, stixObject -> stix_object?

This is largely untested, I wanted to get the ball rolling and start a discussion before continuing.

### CODE REVIEW

**Comment by daemitus** [comment]
Also, logging feels like an utter disaster with everything going through the api class logging methods. If anything, the helper can offer that for the connectors, but everything from the library should probably just initialize its own logger. At least then we'd have a better idea from the logger name as to where something broke, compared to the root logger.

**Comment by daemitus** [comment]
Should things that silently fail and return none, throw instead?
e.g. ExternalReference.generate_id returns None in certain circumstances.

File: pycti/api/opencti_api_client.py [hunk 1]
```diff
@@ -440,126 +441,107 @@ def not_empty(self, value):
...
-            data["edges"] if "edges" in data and data["edges"] is not None else []
-        ):
-            row = edge["node"]
-            if with_pagination:
-                result["entities"].append(self.process_multiple_fields(row))
-            else:
-                result.append(self.process_multiple_fields(row))
-        if with_pagination and "pageInfo" in data:
-            result["pagination"] = data["pageInfo"]
+            edges = data.get("edges") or []
```
Comment (nor3th) [inline pycti/api/opencti_api_client.py:469]:
```suggestion
            edges = data.get("edges", [])
```
Also possible like this ;)

Comment (daemitus) [reply pycti/api/opencti_api_client.py:469]: if for whatever reason there's an edges=None in the mapping, it'll explode. Now that shouldn't happen, but you never know sometimes :p

**Comment by daemitus** [comment]
Is there a reason various methods aren't exposed, while a GQL query exists? Like entities.vulnerability.delete(id) for example

**Review by nor3th — COMMENTED** [review]
Hey @daemitus 

Thank you very much for your great work. I haven't been able to go through everything yet, but I will in the upcoming week.

Please run the tests for your PR, as the CircleCi tests are failing due to python 3.7 not yet supporting and I will have a look if we still need to support 3.7. https://app.circleci.com/pipelines/github/OpenCTI-Platform/client-python/1081/workflows/e4cb5527-febf-4449-8c22-598578dafda0/jobs/3409

Concerning the logging, I fully agree with what you said. The connector library which is currently in development is choosing such an approach, which we could definitely integrate into the rest of the pycti library.
Here's my fork for the connector development: https://github.com/nor3th/client-python/tree/connector_refactor/pycti/connector/new

Regards

**Comment by daemitus** [comment]
> Here's my fork for the connector development

[This](https://github.com/OpenCTI-Platform/connectors/blob/master/external-import/urlscan/src/urlscan/loop.py
) is the core loop ive been using for external imports for any connectors being developed on work time. A uniform way of handling all the exit conditions, ping/run_and_terminate, is a nice thing to have.

**Comment by daemitus** [comment]
Got past the import errors, only to find the demo instance down

**Comment by daemitus** [comment]
threatactor.create seems to be missing `roles`. Is this intended?

**Comment by nor3th** [comment]
Hey @daemitus 

We are aware, that the current code pattern of the pycti entities is not ideal and should be rewritten in a more modular way, but the connector library is more important to rewrite for now. The pycti entities are not designed in a best practice way, but they work alright for now. The idea we are having for the future of pycti is to design the entities in a classic composite design pattern where ie. the ThreatActor class uses a lot of functionality from it's parent class SDO. pydantic might be a great library for this approach, however we are open for any other ideas/approaches. Automatic code documentation based on the function's comment is also on the todo list.

If you already want to improve pycti in a more comprehensive matter, feel free do to so :blush: 

> Should import_from_stix2 be typed as Union[Dict[str, any], stix2.SomeStixType]?

Hmm, give it a try. I think the argument should be a serialized version of the object's stix2.Type. Since typing isn't impacting the functionality, we can always adapt it if it doesn't represent what is actually being used as argument for this function.

> Is there a reason various methods aren't exposed, while a GQL query exists? Like entities.vulnerability.delete(id) for example

Not really, since maintaining every entity separately it is quite easy to lose the overview of what exists and what doesn't. I guess the reason for this is that the `delete` function wasn't needed and thus we forgot to implement it.

> threatactor.create seems to be missing roles. Is this intended?

No, same reason as above. Lost sight of what exists and what doesn't. Please add it.

> Should things that silently fail and return none, throw instead? e.g. ExternalReference.generate_id returns None in certain circumstances.

If they fail, they should throw an exception as None doesn't indicate something went wrong.

> logging

We could simply use this function for pycti logging in general. What do you think? https://github.com/nor3th/client-python/blob/connector_refactor/pycti/connector/new/libs/connector_utils.py#L24-L36

> Deprecating non-snake-case arguments, stixObject -> stix_object?

YES!!!

> Demo instance down.

I'll rerun the tests once it's back up again. 

Open for any ideas, proposals or feedback from your side :blush: 

Regards

**Comment by SamuelHassine** [comment]
Demo is back online, running testing again.

**Comment by SamuelHassine** [comment]
Tests are still broken @nor3th @daemitus. I would like to thank you both for this amazing work. FYI, next release is scheduled for August 31, it would be great if we can have those enhancements, but definitely take the time you need to properly finish them.

**Comment by nor3th** [comment]
Hey @SamuelHassine 

As @daemitus described it, this is PR is more aimed at the code quality rather than real functionality improvements. Unless there is a real demand for having those changes integrated into the upcoming minor release, I'd prefer finishing the discussion and improving the code quality for the release after the upcoming one. Works for you?

Regards

**Comment by daemitus** [comment]
I'm playing around with a code generator to potentially save the trouble of
manually doing 20 more SDOs as well.

**Comment by SamuelHassine** [comment]
No problem @nor3th thanks a lot!

**Comment by nor3th** [comment]
Hey @daemitus

Any updates on your approach?

**Comment by SamuelHassine** [comment]
Closing this for inactivity.

**Pull request closed without merging**